### DISCUSSION TITLE
page_navigate not working. Possible bug?

### DISCUSSION THREAD

**Discussion by OzGav** [ExperienceLovelace/ha-floorplan#181] [Bugs] [answered]
I have recently come up with a use case for wanting to be able to click on a SVG and have that open a new browser window.

I have tried many options but nothing seems to work. In my old reference guide there were floorplan calls of` page_navigate` and `window_navigate`. Looking at the current code I only see `page_navigate` in there (and I found a message from @pkozul saying he removed window_navigate) so I have been trying that. However, the docs also say I can use all HA services so what I think should work is

```
                  tap_action:
                    action: url
#                    service: floorplan.page_navigate
                    url_path: 'http://192.168.1.254'
```

but that throws this error:
`http://192.168.1.165:8123/hacsfiles/ha-floorplan/floorplan.js:41:190 Uncaught SecurityError: Failed to execute 'pushState' on 'History': A history state object with URL 'http://192.168.1.254/' cannot be created in a document with origin 'http://192.168.1.165:8123' and URL 'http://192.168.1.165:8123/lovelace/10'.`

In the action line I have tried url, navigate and call-service
I have tried the `service `line in and out
I have tried `url_path`, `service_data` and `data` for the last line.

In just about every combination I have tried nothing happens on the click. No message on screen, nothing in the floorplan log and nothing inthe HA logs. 

If I use the above with a URL outside of HA then I get that error. If I use `url_path: '/lovelace/9' `then that works.

Any ideas?

**Comment by OzGav** [comment]
@exetico any ideas?

**Comment by exetico** [comment]
You're asking for a function to navigate **out** of Home Assistant, of the **current window** where you have Home Assistant running, right?

As you've found out, `page_navigate` are only made for internal page-navigation. It's actually code from the `home-assistant` repo:
https://github.com/home-assistant/frontend/blob/dev/src/common/navigate.ts
_I bet here's a few changes in the Home Assistant source-code, but... The functions are still almost 1:1._

Our function are simular to [navigation_path](https://www.home-assistant.io/lovelace/actions/#navigation_path) in Home Assistant, which are only made for navigation around the lovelace UI.

There **is** a `url_path` function in Home Assistant, but that's not used in ha-floorplan right now.

Here's a simple example where `url` are in use:
```
      - type: button
        tap_action:
          action: url
          url_path: 'http://192.168.1.2/myotherurl'
```

It's generally bad practice to allow navigation to another URL in a env. like this, and I bet that's why Home Assistant are not allowing that as pr. default.

Are you primary need to open a new window, or change current window, to another external page? In Home Assistant, you only have the option to open a new window with `url` and `url_path` as the service_data. There's `window.open` in use.
https://github.com/home-assistant/frontend/blob/08863348dc64d826dc038b6193e7c31b42d04057/src/panels/lovelace/common/handle-action.ts#L116

We could just extend the functions and allow the usage of `window.location.replace(url_path);`, but I bet there's a reason why Home Assistant are not allowing that in the button card.

However, it is `wrong` that we're just using `navigate()` for the `url` service. It should as least use `window.open`. 

You could consider to ask the Home Assistant dev's (on GitHub) if there's a specific reason why window.location.replace are not allowed/supported as we speak. 

I'm not sure if HA are normally limiting the amount of `property`-changes in the users browsers. The `window.open` are a method. 

With that said, `window.open("http://192.168.1.2/myotherurl","_self")` should do the trick just fine.. So maybe that's the way to fix it - by adding a option to force `_self` as the [windowName](https://developer.mozilla.org/en-US/docs/Web/API/Window/open). 

As you can hear, I'm not that into the security-perspectives, if there's anything we should consider before implementing it.

**Reply by exetico** [reply]
**TL;DR of the long post**:
I would like us to stick to the default way of using Home Assistant, while using the `navigate` and `url` service. However, the current `url`service needs to be fixed. It's currently using `navigate()` instead of `window.open()` - so we're actually not anything near 1:1 with the Home Assistant way of doing it. We *'could** add an option to force `_self` as the `windowname`, but I'm not sure if that a proper solution.
Another way are just to add another option, allowing the user to execute whatever javascript they want, without packing it within a service-call, like `floorplan.execute` or simular. I'm not really sure there. Yet another solution would be to add an `floorplan.none` or simular named function, and just `break;` it with nothing it it. There's a good amount of ways...

**Reply by OzGav** [reply]
> I would like us to stick to the default way of using Home Assistant,

That is what I think I am trying to do. HA actions https://www.home-assistant.io/lovelace/actions/ states there are two actions relevant here. `navigate` and `url`. They come with the data variables of `navigation_path` and `url_path` respectively. The example in the docs shows an external url so I believe I should be using that as you suggested.

@pkozul has in the docs that 
> Floorplan can perform any of the available Lovelace actions 

so based on that I thought it should work. Although you say above that `url `is not used right now? 

Last May @pkozul told me on Discord that he had 
> removed the floorplan.window_navigate as it's redundant... so now, we can just rely on the navigate action from Lovelace

 but I saw that `page_navigate` is still in the code so I thought that might work?

I am happy to use the hack you have provided below but just thought that maybe something that should be working isn't.

P.S. Thanks for the thanks! Happy to help!

**Reply by exetico** [reply]
I think you've missed something in my reply 😄 ...

I've mentioned that the `navigate` function from Home Assistant itself, are not allowing external pages to be opened not navigated in same tab. So we need to decide what to do next. I understand that @pkozul  has told you that it's removed - but that's for a reason. 

Now we need to decide, how to do things in the future. I would personally like to just fix the `url` part so that's aligned with Home Assistant. But with that said, we need to device what should be possible in our "own layer" of functions 😄 ...

Do you have a need to direct navigation from the same tab, or are you okay with the functions just opening a new tab? That's the big question here.

**Reply by OzGav** [reply]
Actually it’s a new tab I am after.

Isn’t the `navigate` function for internal pages and `url` for external?

So for my education is it a bug that the HA actions are not working as the docs say they are supposed to? Or are you saying it is expected as each HA action has to be implemented in floorplan specifically?

i think the reason I am confused is because Petar said he removed window_navigate because the HA Action can be used (ie the `navigate` action) so I assumed page_navigate was for external URLs but that isn’t working? Is it a bug?

**Comment by exetico** [comment]
With that said.. I guess it's a bit dump to talk about security, now that the ha-floorplan user are allowed to execute native JavaScript in the browser, so here's an ugly hack:

```
    action: call-service
    service: |
      > {
        window.open("https://google.com","_self");
        return 'none.none';
      }
```

You would need to add the [empty-service](https://github.com/chomupashchuk/empty-service-home-assistant), for this to work without an error.

It's a bit ugly... But it will work.

ALSO: Thank you for spending your time with all the users in the Discussion, @OzGav! It's almost the most important thing in this project... allowing the user-base to grow, day by day 🥳 !

**Comment by exetico** [comment, marked as answer]
https://github.com/ExperienceLovelace/ha-floorplan/pull/185 now pushed to master-branch. https://github.com/ExperienceLovelace/ha-floorplan/releases/tag/1.0.30 are now ready, too.
